src/coreclr/jit/lowerxarch.cpp

```diff
@@ -6977,6 +7114,53 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                             break;
                         }
 
+                        case NI_SSE41_Insert:
+                        {
+                            GenTree* lastOp = node->Op(numArgs);
+
+                            if ((simdBaseType == TYP_FLOAT) && lastOp->IsCnsIntOrI())
+                            {
+                                // Sse41.Insert has:
+                                //  * Bits 0-3: zmask
+                                //  * Bits 4-5: count_d
+                                //  * Bits 6-7: count_s (register form only)
+                                //
+                                // Where zmask specifies which elements to zero
+                                // Where count_d specifies the destination index the value is being inserted to
+                                // Where count_s specifies the source index of the value being inserted
+
+                                ssize_t ival = lastOp->AsIntConCommon()->IconValue();
+
+                                ssize_t zmask   = (ival & 0x0F);
+                                ssize_t count_d = (ival & 0x30) >> 4;
+                                ssize_t count_s = (ival & 0xC0) >> 6;
+
+                                if (op1->IsVectorZero())
+                                {
+                                    // When op1 is zero, we can contain op1 and modify the mask
+                                    // to zero everything except for the element we're inserting to
+
+                                    MakeSrcContained(node, op1);
+
+                                    zmask |= ~(1 << count_d);
+                                    zmask &= 0x0F;
+
+                                    ival = (count_s << 6) | (count_d << 4) | (zmask);
+                                    lastOp->AsIntConCommon()->SetIconValue(ival);
+                                }
+                            }
+
+                            if (TryGetContainableHWIntrinsicOp(node, &op2, &supportsRegOptional))
+                            {
+                                MakeSrcContained(node, op2);
+                            }
+                            else if (supportsRegOptional)
+                            {
+                                op2->SetRegOptional();
```

Comment:
I've seen this pattern of using `SetRegOptional` if a node cannot be contained - what is the difference between using `SetRegOptional `and marking the node as contained?

Reply:
"RegOptional" (aka "register optional") means the node _may_ end up effectively contained (used from memory: either a local, or a spill temp). Generally, all operations that support memory containment also support `RegOptional`.

That said, this particular pattern is not correct. `RegOptional` has the same use semantics as containment, i. e. it can only be set if we know the node is safe to contain (in the `IsSafeToContainMem` sense).

Reply:
> That said, this particular pattern is not correct. RegOptional has the same use semantics as containment, i. e. it can only be set if we know the node is safe to contain (in the IsSafeToContainMem sense).

That should be getting appropriately checked as part of `TryGetContainableHWIntrinsicOp`. If it isn't, its a fix in that method and not a fix here.

Reply:
> If it isn't, its a fix in that method and not a fix here.

Yes, and I do see it isn't checked appropriately in `TryGetContainableHWIntrinsicOp` (it is always set for `supportsGeneralLoads`).

(Was commenting on that mostly because there is a common alternation of this: `if (IsContainableMemOp() && IsSafeToContainMem(...) { MakeSrcContained(); } else { SetRegOptional() }`)

Reply:
Since this is a more general pre-existing issue, I'll get a fix up in another PR. Should be straightforward since it just requires checking `IsSafeToContainMem` before setting `supportsRegOptional`

Reply:
Fix up https://github.com/dotnet/runtime/pull/77895